analyzers/src/SonarAnalyzer.Common/Rules/ExpressionComplexityBase.cs

```diff
@@ -43,26 +48,36 @@ public abstract class ExpressionComplexityBase<TExpression> : ParameterLoadingDi
             rule = Language.CreateDescriptor(DiagnosticId, MessageFormat, isEnabledByDefault: false);
 
         protected sealed override void Initialize(ParameterLoadingAnalysisContext context) =>
-            context.RegisterSyntaxTreeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
                 {
-                    var root = c.Tree.GetRoot();
-                    var rootExpressions = NoncompoundSubexpressions(root);
-                    var compoundExpressionsDescendants = root.DescendantNodes().Where(IsCompoundExpression).SelectMany(NoncompoundSubexpressions);
-
-                    foreach (var expression in rootExpressions.Concat(compoundExpressionsDescendants))
+                    if (c.Node.Parent?.Kind() is TSyntaxKind kind && (ComplexityIncreasingKinds.Contains(kind) || TransparentKinds.Contains(kind)))
+                    {
+                        return;
+                    }
+                    var complexity = CalculateComplexity(c.Node);
+                    if (complexity > Maximum)
                     {
-                        var complexity = expression.DescendantNodesAndSelf(x => !IsCompoundExpression(x)).Count(IsComplexityIncreasingKind);
-                        if (complexity > Maximum)
-                        {
-                            c.ReportIssue(Diagnostic.Create(rule, expression.GetLocation(), Maximum, complexity));
-                        }
+                        c.ReportIssue(Diagnostic.Create(rule, c.Node.GetLocation(), Maximum, complexity));
                     }
-                });
+                }, ComplexityIncreasingKinds.Concat(TransparentKinds).ToArray());
 
-        private IEnumerable<SyntaxNode> NoncompoundSubexpressions(SyntaxNode node) =>
-            node.DescendantNodes(x => !IsExpressionOrPatternRoot(x) || x == node).Where(x => IsExpressionOrPatternRoot(x) && !IsCompoundExpression(x));
+        private int CalculateComplexity(SyntaxNode node)
+        {
+            return CalculateComplexityRec(node, 0);
+
+            int CalculateComplexityRec(SyntaxNode node, int currentComplexity)
```

Comment:
static?

Reply:
It accesses ExpressionChildren and ComplexityIncreasingKinds so it's not possible, unfortunately.